src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs

```diff
@@ -61,28 +62,43 @@ public TestEngine() : this(TestRuntimeProviderManager.Instance, new ProcessHelpe
+        // has no preference or the preference is 0 then we use the number of logical processors. Or the number of sources, whatever is lower.
+        // We don't know for sure if we will start that many processes as some of the sources can run in a single testhost. This is determined by
+        // Shared on the test runtime provider. At this point we need to know only if the parallel level is more than 1, and so if we will do parallel
+        // run or not.
         var parallelLevel = VerifyParallelSettingAndCalculateParallelLevel(
             discoveryCriteria.Sources.Count(),
             discoveryCriteria.RunSettings);
 
+        var isParallelRun = parallelLevel > 1;
+
         // Collecting IsParallel enabled.
-        requestData.MetricsCollection.Add(
-            TelemetryDataConstants.ParallelEnabledDuringDiscovery,
-            parallelLevel > 1 ? "True" : "False");
-        requestData.MetricsCollection.Add(
-            TelemetryDataConstants.TestSessionId,
-            discoveryCriteria.TestSessionInfo?.Id.ToString() ?? string.Empty);
+        requestData.MetricsCollection.Add(TelemetryDataConstants.ParallelEnabledDuringDiscovery, isParallelRun ? "True" : "False");
+        requestData.MetricsCollection.Add(TelemetryDataConstants.TestSessionId, discoveryCriteria.TestSessionInfo?.Id.ToString() ?? string.Empty);
+
+        // Get testhost managers by configuration, and either use it for in-process run. or for single source run.
+        List<TestRuntimeProviderInfo> testHostManagers = GetTestRuntimeProvidersForUniqueConfigurations(discoveryCriteria.RunSettings, sourceToSourceDetailMap, out ITestRuntimeProvider testHostManager);
 
-        if (ShouldRunInNoIsolation(discoveryCriteria.RunSettings, parallelLevel > 1, false))
+        // This is a big if that figures out if we can run in process. In process run is very restricted, it is non-parallel run
+        // that has the same target framework as the current process, and it also must not be running in DesignMode (server mode / under IDE)
+        // and more conditions. In all other cases we run in a separate testhost process.
+        if (ShouldRunInProcess(discoveryCriteria.RunSettings, isParallelRun, isDataCollectorEnabled: false, testHostManagers))
         {
+            // We are running in process, so whatever the architecture and framework that was figured out is, it must be compatible. If we have more
+            // changes that we want to to runsettings in the future, based on SourceDetail then it will depend on those details. But in general
+            // we will have to check that all source details are the same. Otherwise we for sure cannot run in process.
+            // E.g. if we get list of sources where one of them has different architecture we for sure cannot run in process, because the current
+            // process can handle only single runsettings.
+            var testHostManagerInfo = testHostManagers.Single();
```

Comment:
I would favor a manual check and throw an exception with actionable message. If this happens we will most likely have trouble to fix it.

Reply:
Replaced with a custom exception. I don't expect to ever see it though so I am not localizing. :)